Include CSRF tokens when using the password provider

ProbablePrime opened the pull request.

MCPH-Jenkins commented:
@ProbablePrime, thanks for your PR! By analyzing the blame information on this pull request, we identified @connor4312 and @SimonSchick to be potential reviewers

connor4312 commented on lib/providers/password.js:
```diff
@@ -16,31 +16,46 @@ var errors = require('../errors');
-            return res;
         }
-    });
+        this.setCSRFToken(res.headers[this.csrfTokenLocation]);
```
This isn't necessary. `this.client.request` delegates to `.getRequest` automatically

SimonSchick commented on lib/providers/password.js:
```diff
@@ -16,31 +16,46 @@ var errors = require('../errors');
     this.jar = request.jar();
+    // storage of the CSRF Token as given in the response from the server
+    this.csrfToken = null;
+    this.csrfTokenLocation = 'x-csrf-token';
```
Why store this in each instance? I feel like you can hardcode this or make it a constant (var top of file) instead.

SimonSchick replied:
Would also save some extra code down below 😛

ProbablePrime replied:
I felt that something that didn't have a spec deserved a place to change this header once and have all places update. I don't like hardcoding things like this.

connor4312 replied:
To be clear the update document is a spec of our implementation of CSRF cookies. If we introduce a change to the header, it would probably be a behavioural change and not just a renaming.

ProbablePrime replied:
Ill bump it to a var, i still don't want to hard code it.

connor4312 commented:
The ideal behaviour with CSRF is that if we get a 401 CSRF on _any_ request, we update the stored CSRF token and retry the request. CSRF tokens may expire at any time.

ProbablePrime commented:
This wasn't specified in the changes document that they expire.  Will address :)

connor4312 commented:
Ah, that was an oversight on my part then, I'll adjust that

connor4312 commented:
update, it is included 😄 

>  They are invalidated after the session expires and may be invalidated at any point prior to the session expiration.

 I'll emphasize it a bit more though

ProbablePrime commented:
Wasn't quite sure how to handle automatic retries. Implemented one way, let me know what you think.

Oh also refactored a bit, provider is now a parent class of password and oauth :)

connor4312 commented on lib/providers/password.js:
```diff
@@ -14,33 +15,53 @@ var errors = require('../errors');
+    return req;
+};
+
+PasswordProvider.prototype.handleResponseError = function (err) {
```
Perhaps, rather than returning a boolean from this (which isn't super obvious behaviour), it optionally returns a promise. Then we can have it just call `this.client.request` and return from there. A bit less single-purpose structure than what it is currently.

ProbablePrime replied:
So pass the arguments of the request down to the provider and pas those to the call to .request inside the provider?

connor4312 replied:
I think that would be a reasonable approach, `(err: Error, options: RequestOptions) => Promise<any>?`

connor4312 commented:
One cleanup comment. Looks good overall.

SimonSchick commented on lib/client.js:
```diff
@@ -128,7 +128,13 @@ Client.prototype.request = function (method, path, data) {
             resolve(res);
         });
-    }).bind(this);
+    }).bind(this)
```
Are you sure this works without bluebird?

ProbablePrime replied:
bluebird is explicitly used in this method. When would bluebird not exist?

SimonSchick replied:
If I remember correctly bluebird is ignored in the frontend build and replaced with a generic promise implementation, @connor4312 isn't that correct?

connor4312 replied:
Current

ProbablePrime replied:
We should make this more transparent. 
```
const Beam = require('beam-client-node');
Beam.Promise = <Promise constructor you want>
```
Then `Beam.Promise` in place of Bluebird.

Or something.

Right now that's super unclear to me. As a short term i'll test this with the frontend promise provider.

connor4312 replied:
I'm not sure that `beam-client-node` is actually used in the frontend at the moment. That may change in the future as we move off wsabi

ProbablePrime commented:
Ahh, updating typings  i forgot them

ProbablePrime commented:
@UnwrittenFun Typings review please.

UnwrittenFun commented on lib/providers/oauth.d.ts:
```diff
@@ -2,8 +2,9 @@ import Client = require("../client");
 
 import { BeamRequest } from "../../defs/request";
 import { OAuthToken } from "../../defs/oauth";
+import Provider = require("./provider");
```
wat

UnwrittenFun commented on lib/providers/password.d.ts:
```diff
@@ -5,12 +5,21 @@ import Client = require("../client");
   jar: request.CookieJar;
   /**
+   * Storage of the CSRF Token as given in the response from the server
+   * @type {String}
```
rm type

UnwrittenFun commented on lib/providers/provider.d.ts:
```diff
@@ -0,0 +1,32 @@
+  /**
+   * Given a failing response from the client, processes the error object and returns a Promise
+   * which allows for a provider to retry a request or carry out some other process
+   * @return {Promise.<>}
```
rm return

UnwrittenFun commented on lib/providers/provider.d.ts:
```diff
@@ -0,0 +1,32 @@
+  /**
+   * Returns info to add to the client's request.
+   */
+  getRequest(): Object;
```
any

UnwrittenFun commented on lib/providers/provider.d.ts:
```diff
@@ -0,0 +1,32 @@
+   * which allows for a provider to retry a request or carry out some other process
+   * @return {Promise.<>}
+   */
+  handleResponseError(): Promise<any>;
```
Comment seems to suggest `Promise<void>` ?

ProbablePrime commented:
@connor4312 One more review please.

connor4312 commented:
lgtm
